### DISCUSSION TITLE
C++ Raw Window Handle

### DISCUSSION THREAD

**Discussion by shakesbeare** [slint-ui/slint#6248] [Q&A]
Relevant: https://github.com/slint-ui/slint/discussions/5107

I'm working on a VST plugin host project. Since the SDK is written in C++, it's been quite difficult to get it off the ground in Rust. Part of what I need to do is open a blank window and provide the plugin with a pointer to said window so it can draw it's editor. In Rust with native `winit`, I could basically just grab the `RawWindowHandle` and transform it into a `*mut c_void`. 

From the linked discussion, it seems the same solution is not available using Slint in C++ as it isn't yet exposed to the C++ API. But I also am wondering if this is even the best way to go about it in a framework like Slint. Will Slint be trying to compete with the plugin to render onto the window? 

We'd like to use Slint as a solution for our general app interface, but also have this additional requirement of drawing editors for VST plugins. My limited testing has seemingly ruled out using native windows for this as `winit` doesn't like to be run on a non-principal thread and (at least on MacOS) the native frameworks also have this preference. So, it makes sense to try to use Slint to manage both sorts of windows.

Another idea I've had is to make "virtual windows" of sorts -- some kind of sub-surface that the plugin will treat as a window but which only exists inside of the Slint ecosystem. Is this doable? 

Any feedback would be appreciated.

**Comment by ogoffart** [comment]
Right, we should add API to get the raw window handle as a void* in C++ 

Do I understand correctly that you're not using Slint do implement the VST plugin, but instead, want to display VST plugins from a Slint application?

But there is still the option of using native windows without winit as Slint can be drawn on native window with the [Slint Platform API](
https://releases.slint.dev/1.7.2/docs/cpp/api/namespace_slint__platform)
Example for Windows https://github.com/slint-ui/slint/blob/master/examples/cpp/platform_native/windowadapter_win.h , (we have similar code for MacOS but it hasn't been pushed into an example.)
Maybe that can be the solution.


We have a [class NativeWindowHandle](https://releases.slint.dev/1.7.2/docs/cpp/api/classslint_1_1platform_1_1nativewindowhandle#class-nativewindowhandle) in our C++ API, but it is used to give the pointer from the application, to Slint. Not the other way around.
If we were to implement that in C++, we'd need to get the handle.

**Reply by shakesbeare** [reply]
> Do I understand correctly that you're not using Slint do implement the VST plugin, but instead, want to display VST plugins from a Slint application?

Yes that's correct. 

> But there is still the option of using native windows without winit as Slint can be drawn on native window 

This seems like a good approach, I hadn’t seen that. Our goal was to make a cross platform solution so we’d prefer not to have to handle the windows natively where possible, but it’s acceptable. 

I’d be happy to work on getting the API updated. Although, I was struggling to get my bearings looking through that bit of the code and I might need  a fair amount of guidance. Otherwise, consider this a +1 for exposing the raw handles to the C++ API. 

Thanks for your response!

**Comment by ogoffart** [comment]
If you were to implement the raw window handle for C++, you'd need to add a getter in the Rust FFI. (as an example, you can see the getter for the position) https://github.com/slint-ui/slint/blob/master/internal/core/window.rs#L1427
In the getter you can use slint::Window::window_handle().window_handle()/display_handle() and somehow extract the pointer and their kind.
Then we'd need to figure a public C++ API in Window slint_window.h
Perhaps
```c++
class Window { 
   //...
   std::optional<std::tuple<HWND, HINSTANCE>> win32_handle() const;
   std::optional<std::tuple<NSWindow*, NSVIew*>> appkit_handle() const;
   // ...
```
Or maybe std::tuple is not a great API and we need a more specific struct, 
or could also be:
```c++
   bool win32_handle(HWND &hwnd, HINSTANCE &hinstance) const;
```

Not sure what is best, none of these options sounds super great to me.

**Reply by marlonmarcello** [reply]
This doesn't seem so bad to me. Would love to hear your concerns.

I've been working on a custom desktop shell targetting Wayland and Hyprland and got to a point where I wanted to integrate a UI. I was messing around with Slint and having a lot of fun with it and wanted to see if I could use it to also be the taskbar, but to render it as a layer I need access to the window handle. The project is already in C++.
